### PR TITLE
8319773: Avoid inflating monitors when installing hash codes for LM_LIGHTWEIGHT

### PR DESCRIPTION
**Pull request by xmas92**
LM_LIGHTWEIGHT only uses the lock bits for its locking. This leaves the hashCode bits free when a monitor is not inflated. So instead of inflating when installing the hashCode on a fast locked object it can simply use the hashCode bits in the markWord.

The mark word transitions Unlocked (0b01) <=> Locked (0b00) are done by retrying the CAS if it fails due to non-lock bit changes. 
The mark word transitions Monitor (0b10) <=> Locked/Unlocked (0b0X) are the same as before, inflation already handles hash codes. This change does not interact with the mark word if it is in a Monitor (0b10) state, so the strong CAS which is used for deflation are still valid, and will not fail to any other reason than the cooperative race to help transition the mark word during deflation.

This is dependent on JDK-8319778 simply because JDK-8319797 is dependent on both this and JDK-8319778.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319773](https://bugs.openjdk.org/browse/JDK-8319773): Avoid inflating monitors when installing hash codes for LM_LIGHTWEIGHT (**Enhancement** - P4)

### Reviewers
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to [46d4c1b1](https://git.openjdk.org/jdk/pull/16603/files/46d4c1b178c129db6ced0807e7dbcc345b06e859)
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [fdbfbf8a](https://git.openjdk.org/jdk/pull/16603/files/fdbfbf8a9548f1e20fe076e385b2f102652134c8)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16603/head:pull/16603` \
`$ git checkout pull/16603`

Update a local copy of the PR: \
`$ git checkout pull/16603` \
`$ git pull https://git.openjdk.org/jdk.git pull/16603/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16603`

View PR using the GUI difftool: \
`$ git pr show -t 16603`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16603.diff">https://git.openjdk.org/jdk/pull/16603.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16603#issuecomment-1805644121)

### CODE REVIEW

**Comment by xmas92** [comment]
## Testing
  - `linux-x64`, `linux-x64-debug`
    - `LM_LEGACY`
      - [X] tier 1-7
    - `LM_LIGHTWEIGHT`
      - [ ] tier 1-7
  - [X] GitHub actions

**Review by rkennke — APPROVED** [review]
This is great! I've done something like that in my original LW-locking PR, but then ripped it out to keep it simpler. Never got around to re-do it. Changes look good to me!

File: src/hotspot/share/runtime/synchronizer.cpp [hunk 1]
```diff
@@ -572,23 +573,14 @@ void ObjectSynchronizer::exit(oop object, BasicLock* lock, JavaThread* current)
...
     if (LockingMode == LM_LIGHTWEIGHT) {
       // Fast-locking does not use the 'lock' argument.
-      if (mark.is_fast_locked()) {
-        markWord unlocked_mark = mark.set_unlocked();
-        markWord old_mark = object->cas_set_mark(unlocked_mark, mark);
-        if (old_mark != mark) {
-          // Another thread won the CAS, it must have inflated the monitor.
-          // It can only have installed an anonymously locked monitor at this point.
-          // Fetch that monitor, set owner correctly to this thread, and
-          // exit it (allowing waiting threads to enter).
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/synchronizer.cpp:582]: I don't understand why the anonymous owner case is no longer being checked. ??

File: test/hotspot/jtreg/runtime/whitebox/TestWBDeflateIdleMonitors.java [hunk 2]
```diff
@@ -59,14 +59,22 @@ public static void main(String args[]) throws Exception {
...
         public static Object obj;
 
-        public static void main(String args[]) {
+        public static void main(String args[]) throws Exception {
             obj = new Object();
             synchronized (obj) {
-                // HotSpot implementation detail: asking for the hash code
-                // when the object is locked causes monitor inflation.
-                if (obj.hashCode() == 0xBAD) System.out.println("!");
+                if (LockingMode != LM_LIGHTWEIGHT) {
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/whitebox/TestWBDeflateIdleMonitors.java:69]: I don't think we need to make this test behave differently because of implementation details of different locking modes. Just use `wait(1)` unconditionally. Thanks.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Seems reasonable - though more complex than I had envisaged. One query below and a change requested for the test.

Thanks

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 1 above)
Comment (xmas92) [reply src/hotspot/share/runtime/synchronizer.cpp:582]: The condition does now check for a successful CAS, not the unsuccessful one. If it was successful then there is no monitor, thus no anonymous owner.

If the CAS failed and the mark word is no longer fast locked. It must be inflated. So we fallthrough down to the inflated case.

`ObjectSynchronizer::inflate` correctly handles fixing the owner.

File: test/hotspot/jtreg/runtime/whitebox/TestWBDeflateIdleMonitors.java (hunk 2 above)
Comment (xmas92) [reply test/hotspot/jtreg/runtime/whitebox/TestWBDeflateIdleMonitors.java]: Done.

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/runtime/synchronizer.cpp:582]: I see - thanks. It is hard to see where the code goes to when the CAS fails.

**Review by dholmes-ora — COMMENTED** [review]
Noting further from me. Runtime changes look good. Best to get a compiler dev to okay the C2 changes.

Thanks

File: src/hotspot/share/runtime/synchronizer.cpp [hunk 3]
```diff
@@ -963,13 +959,6 @@ intptr_t ObjectSynchronizer::FastHashCode(Thread* current, oop obj) {
       }
       // Fall thru so we only have one place that installs the hash in
       // the ObjectMonitor.
-    } else if (LockingMode == LM_LIGHTWEIGHT && mark.is_fast_locked() && is_lock_owned(current, obj)) {
```
Comment (coleenp) [inline src/hotspot/share/runtime/synchronizer.cpp:966]: You can delete the is_lock_owned() function just above this with this patch.

Comment (coleenp) [reply src/hotspot/share/runtime/synchronizer.cpp:966]: Also I think the comment before this function is wrong.  We can inflate for deoptimization.

Comment (xmas92) [reply src/hotspot/share/runtime/synchronizer.cpp:966]: > but you could move it down to before 'inflate'
Done.

> Also I think the comment before this function is wrong. We can inflate for deoptimization.
The comment on `is_lock_owned`? For relock_objects `current` is the deoptee's thread which is a JavaThread.

Comment (coleenp) [reply src/hotspot/share/runtime/synchronizer.cpp:966]: Yes that comment.  If we don't inflate for hashcode anymore, why would we inflate with a non-JavaThread? Is it impossible?

Comment (xmas92) [reply src/hotspot/share/runtime/synchronizer.cpp:966]: I see what you mean. Because we need to put the hash code in the ObjectMonitor if it already is inflated and it may be the case that the ObjectMontior has an anonymous owner we can still call `is_lock_owned` when retrieving the monitor for FastHashCode. 

Technically we could condition the inflate call on if the mark word lock bits and pick out the ObjectMontior from the mark word. It would always be installed when reaching that point in the LM_LIGHTWEIGHT mode.

File: src/hotspot/share/runtime/synchronizer.cpp [hunk 4]
```diff
@@ -513,11 +513,12 @@ void ObjectSynchronizer::enter(Handle obj, BasicLock* lock, JavaThread* current)
...
-        if (mark.is_neutral()) {
-          assert(!lock_stack.contains(obj()), "thread must not already hold the lock");
+        while (mark.is_unlocked()) {
           // Try to swing into 'fast-locked' state.
-          markWord locked_mark = mark.set_fast_locked();
-          markWord old_mark = obj()->cas_set_mark(locked_mark, mark);
+          assert(!lock_stack.contains(obj()), "thread must not already hold the lock");
+          const markWord new_mark = mark.set_fast_locked();
+          const markWord old_mark = mark;
+          mark = obj()->cas_set_mark(new_mark, old_mark);
```
[the remark this answers is not in the document]
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/synchronizer.cpp:521]: The change from `mark.is_neutral()` to `mark.is_unlocked()` is equivalent
since both functions have the exact same meaning. Is there some reason to
make that change? Are we trying to migrate away from `mark.is_neutral()`  to
`mark.is_unlocked()`? I think the original "neutral" concept was from when we
had biased locking...

[the remark this answers is not in the document]
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/synchronizer.cpp:521]: The rename from `locked_mark` to `new_mark` seems superfluous. The old and
new version of the variable name are used in the same places so I don't quite see
the reason for the rename. I liked `locked_mark` because it was the "locked" version
of the original `mark`. I do like the use of the `const`.

[the remark this answers is not in the document]
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/synchronizer.cpp:521]: Now I think I see why you changed the `cas_set_mark()` call to return its value into
`mark` instead of `old_mark` and why the original parameter to the CAS was changed
from `mark` to `old_mark`. It's a bit different style then we typically use for CAS where
the return value goes into an `old_something` variable.

Since `mark` is now the loop control and `old_mark` is now `const` you had to juggle
things with the `cas_set_mark()` call. Okay, I think I get the reason for this bit of juggling.
(I still don't understand the rename from `locked_mark` to `new_mark` but that's okay.)

[the remark this answers is not in the document]
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/synchronizer.cpp:521]: Okay so this bit of code changes were done for a couple of reasons (of course this is my guess):
1) add a retry here in case of collision to give this thread another chance at doing a lightweight lock instead of always dropping into inflate-enter loop below on a collision
2) clean up the code a little bit with some use of `const` to make things a little more clear.

I'm going to guess that the type of collision that we're trying to avoid is hash code installation.

You may want to add a comment above `L516 while (mark.is_unlocked()) {` that says:
```
// Allow for a retry here just in case the cas_set_markI() below collided with a hash code installation:
```

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 1 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/synchronizer.cpp:582]: In the old code, I think if there was a hash code installation racing with this
monitor exit by the thread owner, the hash code installer thread would have
inflated the monitor which would have resulted in the anonymous owner value
also being set. So we could have had two different inflating thread racers...

File: src/hotspot/share/runtime/synchronizer.cpp [hunk 5]
```diff
@@ -572,23 +573,14 @@ void ObjectSynchronizer::exit(oop object, BasicLock* lock, JavaThread* current)
...
-          // Another thread won the CAS, it must have inflated the monitor.
-          // It can only have installed an anonymously locked monitor at this point.
-          // Fetch that monitor, set owner correctly to this thread, and
-          // exit it (allowing waiting threads to enter).
-          assert(old_mark.has_monitor(), "must have monitor");
-          ObjectMonitor* monitor = old_mark.monitor();
-          assert(monitor->is_owner_anonymous(), "must be anonymous owner");
-          monitor->set_owner_from_anonymous(current);
-          monitor->exit(current);
+      while (mark.is_fast_locked()) {
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/synchronizer.cpp:576]: Please consider add a comment above L576:
```
// Allow for a retry here just in case the cas_set_mark() below collided with a hash code installation:
```

File: test/hotspot/jtreg/runtime/whitebox/TestWBDeflateIdleMonitors.java [hunk 6]
```diff
@@ -61,12 +61,11 @@ public static class InflateMonitorsTest {
         static WhiteBox wb = WhiteBox.getWhiteBox();
         public static Object obj;
 
-        public static void main(String args[]) {
+        public static void main(String args[]) throws Exception {
             obj = new Object();
             synchronized (obj) {
-                // HotSpot implementation detail: asking for the hash code
-                // when the object is locked causes monitor inflation.
-                if (obj.hashCode() == 0xBAD) System.out.println("!");
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/runtime/whitebox/TestWBDeflateIdleMonitors.java:69]: Why was this deleted from the test? `get_next_hash()` can still
return this value so it's useful to know if a bad hash code made it
out this far...

File: src/hotspot/cpu/x86/sharedRuntime_x86.cpp [hunk 7]
```diff
@@ -59,9 +60,16 @@ void SharedRuntime::inline_check_hashcode_from_object_header(MacroAssembler* mas
...
 
-  // check if locked
-  __ testptr(result, markWord::unlocked_value);
-  __ jcc(Assembler::zero, slowCase);
+
+  if (LockingMode == LM_LIGHTWEIGHT) {
+    // check if monitor
+    __ testptr(result, markWord::monitor_value);
+    __ jcc(Assembler::notZero, slowCase);
+  } else {
```
Comment (dcubed-ojdk) [inline src/hotspot/cpu/x86/sharedRuntime_x86.cpp:68]: Not needed for other platforms? Or will that be done with other bugs or sub-tasks?

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up. I have a couple of editorial comments, but I think that's it.

I did a review of the C2 changes and I don't see anything obviously wrong,
but having a C2 reviewer would be very, very useful.

This needs additional testing. I recommend Tier{1,2,3} on all the usual Oracle
platforms. Tier{4,5,6} would catch the uses of hash code stuff in more stress
related configs by the Serviceability tests.

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 3 above)
Comment (xmas92) [reply src/hotspot/share/runtime/synchronizer.cpp:966]: LM_LEGACY and LM_MONITOR will. LM_LIGHTWEIGHT technically may. If deflation finishes between reading the mark word in FastHashCode and reading the mark word in `inflate`. It seems like a rare enough case that it does not need to be handled separately. The following change would avoid inflation all together.
```diff
    // An async deflation can race after the inflate() call and before we
    // can update the ObjectMonitor's header with the hash value below.
+   if (LockingMode == LM_LIGHTWEIGHT) {
+     assert(mark.has_monitor(), "must be");
+     monitor =  mark.monitor();
+   } else {
-   monitor = inflate(current, obj, inflate_cause_hash_code);
+     monitor = inflate(current, obj, inflate_cause_hash_code);
+   }
    // Load ObjectMonitor's header/dmw field and see if it has a hash.
```

Maybe I should change it to this. Given that there has been confusion here.
My ideal solution would be to separate the implementations for the different locking modes all together, all of these functions are littered with if (LockingMode == X).

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (xmas92) [reply src/hotspot/share/runtime/synchronizer.cpp:521]: > The change from mark.is_neutral() to mark.is_unlocked() [...]. Is there some reason to make that change?

It made the code more readable for me. The is unlocked conveys some relevant meaning, is natural requires knowledge that an unlocked mark word is neutral. 
The code then reads as transition the mark word from unlocked to to fast locked. Instead of transition the mark word from neutral to fast locked. It may just be my unfamiliarity with the this code ands its history which makes me think the former is more easily understandable then the latter. 

> Are we trying to migrate away from `mark.is_neutral()` to `mark.is_unlocked()`?

There probably is some history here I am unaware of. But to me only one should exist, or the neutral property's meaning needs to be explained in the markWord.hpp file so it is clear what makes it distinct from unlocked. 
If they are just aliases then to me unlocked seems more natural, but could replace all unlocked with neutral as well.

> I still don't understand the rename from `locked_mark` to `new_mark`

I'll rename it. I believe it is coincidental. I removed the old code and implemented the new. `cas_set_mark` calls it parameters `old_mark` and `new_mark`, so I did as well. 

> It's a bit different style then we typically use for CAS where the return value goes into an `old_something` variable.

As this is the case I will switch to that style. To reduce confusion.
```c++
        const markWord locked_mark = mark.set_fast_locked();
        const markWord old_mark = obj()->cas_set_mark(locked_mark, mark);
        if (old_mark == mark) {
          // Successfully fast-locked, push object to lock-stack and return.
          lock_stack.push(obj());
          return;
        }
        mark = old_mark;
```

> You may want to add a comment above `L516 while (mark.is_unlocked()) {` that says:
> 
> ```
> // Allow for a retry here just in case the cas_set_mark() below collided with a hash code installation:
> ```

I'll add a comment. But maybe it should instead say 
```
// Retry until a lock state change has been observed.  cas_set_mark() may collide with non lock bits modifications.
```
This comment should age better when/if we start using more than just the lock bits and hash code bits.

File: src/hotspot/cpu/x86/sharedRuntime_x86.cpp (hunk 7 above)
Comment (xmas92) [reply src/hotspot/cpu/x86/sharedRuntime_x86.cpp:68]: I only found that x86 and 32bit arm does this in the shared runtime. I can create a jbs entry for arm. If it is the case that other platform do this for C1, then I cannot find where.

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/share/runtime/synchronizer.cpp]: > There probably is some history here I am unaware of. But to me only one should exist, or the neutral property's meaning needs to be explained in the markWord.hpp file so it is clear what makes it distinct from unlocked.

There is no distinction and there never has been. The `is_neutral` came in with some backend performance enhancements related to adaptive spinning, around the same time that biased-locking was introduced. The work was done by two different engineers. The code in synchronizer.cpp only used `is_neutral` not `is_unlocked`. But the two functions have always had identical definitions. We should get rid of `is_neutral`, but probably not in this PR.

File: src/hotspot/share/opto/library_call.cpp [hunk 8]
```diff
@@ -4546,14 +4546,22 @@ bool LibraryCallKit::inline_native_hashcode(bool is_virtual, bool is_static) {
...
+    Node *monitor_val   = _gvn.MakeConX(markWord::monitor_value);
+    Node *chk_monitor   = _gvn.transform(new CmpXNode(lmasked_header, monitor_val));
+    Node *test_monitor  = _gvn.transform(new BoolNode(chk_monitor, BoolTest::eq));
 
-  generate_slow_guard(test_unlocked, slow_region);
+    generate_slow_guard(test_monitor, slow_region);
+  } else {
+    Node *unlocked_val      = _gvn.MakeConX(markWord::unlocked_value);
+    Node *chk_unlocked      = _gvn.transform(new CmpXNode( lmasked_header, unlocked_val));
+    Node *test_not_unlocked = _gvn.transform(new BoolNode( chk_unlocked, BoolTest::ne));
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/library_call.cpp:4561]:
```suggestion
    Node *chk_unlocked      = _gvn.transform(new CmpXNode(lmasked_header, unlocked_val));
    Node *test_not_unlocked = _gvn.transform(new BoolNode(chk_unlocked, BoolTest::ne));
```

**Review by TobiHartmann — APPROVED** [review]
Compiler changes look good and in-line with the runtime changes.

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 3 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/synchronizer.cpp:933]: So what did you decide to do here?

**Review by dcubed-ojdk — APPROVED** [review]
I've re-reviewed the changes from v03 -> v08.

Thumbs up with what you currently have. There's still one open query
in `FastHashCode` about whether the VMThread can get in there
during JVM/TI tagging... It may be best to file a follow up bug for
chasing down that detail and possibly addressing it... Or you could
address it with one more change in this PR...

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 3 above)
Comment (xmas92) [reply src/hotspot/share/runtime/synchronizer.cpp:933]: For now I believe the extra code noise from trying to handle this race with deflation is not worth it. I creates some questionable code paths and head scratchers. If we were to add a separate FastHashCode just for LM_LIGHTWEIGHT it would be worth it as the while loop body would look quite a bit different and be easier to reason about. 

But I was looking for input if we should handle this case regardless of code complexity. Or maybe taking this all the way and create a separate FastHashCode with its own more understandable logic which does not have to try to fit in with the legacy locking/inflation protocol.

Regardless if we were to just go with it as it is now there should probably be a comment here along the line:
```c++
  // With LM_LIGHTWEIGHT FastHashCode may race with deflation here and cause a monitor to be re-inflated.
```

Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/synchronizer.cpp:972]: Okay I've re-read this group of comments multiple times and above you wrote:

> Regardless if we were to just go with it as it is now there should probably be a comment here along the line:

```
>   // With LM_LIGHTWEIGHT FastHashCode may race with deflation here and cause a monitor to be re-inflated.
```

To fully understand whether the comment is correct or not, I need to understand
where the "here" is that you want to place this comment. You deleted old line 972
thru old line 978. Is this new comment going to be a replacement for those lines?
Or will the new comment be somewhere else?

**Review by dcubed-ojdk — APPROVED** [review]

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 3 above)
Comment (xmas92) [reply src/hotspot/share/runtime/synchronizer.cpp:972]: Before the call to inflate which is the only control flow path which can observe the race. But it could also be a more general comment overarching the whole FastHashCode. 

The race occurs between reading the mark in FastHashCode and reading the mark in inflate. The ObjectMonitor found in the first mark read may differ from the one returned from inflate, and it may be that the current thread which called FastHashCode is the one that inflates the monitor.

I believe we can leave this enhancement as is, it solves the common case by avoiding inflation when an object is lightweight locked. The race where a re-inflation may occur can be plugged as a separate enhancement.

**Comment by xmas92** [comment]
Ran tier1-5 all Oracle platforms with both LM_LEGACY and LM_LIGHTWEIGHT after the merge. 
Will go ahead and integrate this. I'll create an RFE for the arm32 hashcode optimisation as well as for the re-inflate race.
